File: 19-RemoveNthNodeFromEndOfList/19-RemoveNthNodeFromEndOfList.py
# Last updated: 07/04/2026, 11:43:20
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]: 

        dumy = f = s = ListNode(val = 0 , next = head )

        for i in range(n) :
            s = s.next 

        while s.next : 
            s= s.next 
            f = f.next 

        f.next = f.next.next 

        return dumy.next







        

File: 19-RemoveNthNodeFromEndOfList/19-RemoveNthNodeFromEndOfList.cpp
// Last updated: 07/04/2026, 11:43:11
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        // Create a dummy node to handle edge cases
        ListNode* dummy = new ListNode(0, head);
        ListNode* fast = dummy;
        ListNode* slow = dummy;

        // Move fast n steps ahead
        for (int i = 0; i < n; ++i) {
            fast = fast->next;
        }

        // Move fast to the end, maintaining the gap
        while (fast->next != nullptr) {
            fast = fast->next;
            slow = slow->next;
        }

        // Remove the nth node from end
        slow->next = slow->next->next;

        // Return the updated head
        return dummy->next;
    }
};

